File: frontend/src/services/threadService.ts
import { MESSAGE_PAGE_SIZE } from "@/domain/constants"
import type { HistoryPage } from "@/domain/entities/chat"
import { threadRepository } from "@/repositories/threadRepository"

export async function loadThreadHistory(
  conversationId: string,
): Promise<HistoryPage> {
  return threadRepository.fetchThreadMessages(
    conversationId,
    MESSAGE_PAGE_SIZE,
    0,
  )
}

export async function loadOlderThreadMessages(
  conversationId: string,
  loadedCount: number,
): Promise<HistoryPage> {
  return threadRepository.fetchThreadMessages(
    conversationId,
    MESSAGE_PAGE_SIZE,
    loadedCount,
  )
}

export async function streamThreadTurn(
  conversationId: string,
  userMessage: string,
  onChunk: (chunk: string) => void,
  signal?: AbortSignal,
) {
  return threadRepository.sendThreadMessage(conversationId, userMessage, onChunk, signal)
}

File: ThreadGPTiOSV2/ThreadGPTiOSV2/Application/ThreadService.swift
import Foundation

final class ThreadService {
    private let repository: ThreadRepository

    init(repository: ThreadRepository) {
        self.repository = repository
    }

    func loadThreadHistory(conversationID: String) async throws -> ThreadHistoryPage {
        try await repository.fetchThreadMessages(
            conversationID: conversationID,
            limit: ThreadGPTConstants.messagePageSize,
            offset: 0
        )
    }

    func loadOlderThreadMessages(conversationID: String, loadedCount: Int) async throws -> ThreadHistoryPage {
        try await repository.fetchThreadMessages(
            conversationID: conversationID,
            limit: ThreadGPTConstants.messagePageSize,
            offset: loadedCount
        )
    }

    func streamThreadTurn(
        conversationID: String,
        userMessage: String
    ) -> AsyncThrowingStream<StreamEvent, Error> {
        repository.sendThreadMessage(conversationID: conversationID, userMessage: userMessage)
    }
}

